### ISSUE TITLE
Trigger a flow based on data dependency availability

### ISSUE THREAD

**Issue by chengren311** [azkaban/azkaban#1526]
It is common for a scheduled flow to have data dependencies. With the data availability check feature for Azkaban flows, it will be easier for jobs in an Azkaban workflow to start only when the data they depend on becomes available. 

Motivation
1. Improve user productivity by standardizing how Azkaban flows specify data dependencies. Users don’t have to build their own jobs that check if the data their flows depend on are available.  

2. Reduce overall flow latency and error rate. Some flows are scheduled based on estimated data availability time. The flows will fail when the data is not available when they start. To reduce such failure rates, users will add some buffer time e.g. schedule a flow to start at 10 AM when the dependent dataset is typically available at 8 AM. Despite the buffer, the flow may still fail if the data is delayed for longer than 2 hours.  This feature allows these flows to be scheduled at an earlier time than they currently do so. In the example, the flow can be scheduled to run at 8 AM without worrying out failures due to data availability.  By removing the failures caused by data delays, the feature makes it easier to see true flow logical failures. 

3. Improve overall infrastructure efficiency and stability.  Azkaban can check the data availability efficiently. See the detailed design section below for details. A typical implementation of a custom job that checks for data availability relies on polling HDFS/Hive Metastore. A bad implementation or too many such jobs running concurrently can put pressure on the HDFS namenode/ Hive Metastore. 

4. Encourage users to provide the data dependency information as metadata that Azkaban can understand. Azkaban can leverage this information in the future to gain more flexibility or optimize the system. E.g. schedule the flow in a cluster where the data lives. 
bring the data to the cluster that the flow runs. This can happen if the flow relies on GPU in a certain Hadoop cluster for example.

**Comment by juhoautio** [comment]
👍  for this first of all.

I'm mostly curious about how this will be reflected in the UI and what it will looks like in job configuration. Would you have any examples to share about those?

After all, from my perspective it maybe seems like this interface may not be as useful as it can be limiting when compared to custom polling scripts.. With an explicit job to poll for data it is quite clearly visible from the Job List where the flow is going, how long it has been "blocked by a data dependency", and why (you can log details in the job log of the data polling job). Personally I would really hope to have a similar level of visibility in the UI for the dependency checks.

**Comment by ameyamk** [comment]
Yes - we spent lot of time thinking if data dependency should be modeled as as a job - or should it be designed at a flow level. After lot of thinking we have decided to design this at a flow level. There are couple of reasons for this:

1. We feel data dependency is similar to any existing or future trigger we may have. Such as: schedule based triggers, manual triggers, flow-to-flow dependency trigger etc. While the actual flow is "logic" of the flow. Trigger is logic of how flow is "triggered". This separation of concepts makes it easy to handle entire flow as a simpler concept which can be moved around if needed.

2. This choice does make it harder to display this on UI - compared to if we just made it a simple job type. So we will spend sometime updating UI to reflect some of the key points you made earlier. 

3. One other important change on the radar is - introducing "flow" as a concept at compile time. Today flow is purely "run time" concept. As a result only way to add "flow" level properties is via UI. eg. schedule, SLA etc. We want to take this opportunity and create "flow" as a compile time construct - so that config of the flow such as schedule, data dependancies, sla's can be defined at compile time as well.

**Comment by juhoautio** [comment]
@chengren311 this seems to be in progress - nothing done for UI side yet? What happens to a pending data dependency trigger if azkaban is restarted while it's waiting for data?

**Comment by chengren311** [comment]
@juhoautio yes it's under progress. UI part is still under progress. 
> What happens to a pending data dependency trigger if azkaban is restarted while it's waiting for data?
The pending one will recover from db since all pending dependencies will be persisted into db.
